Add RMI support

pawelchcki opened the pull request:
This PR in its current form adds instrumentation of client and server endpoints used in Java RMI communication.
Both new RMI dynamically generated communication stubs and `rmic` generated Stubs are supported.

```
 * Main entry point for transferring context between RMI service.
 *
 * <p>We into StreamRemoteCall constructor used for invoking remote tasks and performs a
 * backwards compatible check to ensure if the other side is prepared to receive context propagation
 * messages then if successful sends a context propagation message
 *
 * <p>Context propagation consist of a Serialized HashMap with all data set by usual context
 * injection, which includes things like sampling priority, trace and parent id
 *
 * <p>As well as optional baggage items
 *
 * <p>On the other side of the communication a special Dispatcher is created when a message with
 * DD_CONTEXT_CALL_ID is received.
 *
 * <p>If the server is not instrumented first call will gracefully fail just like any other unknown
 * call. With small caveat that this first call needs to *not* have any parameters, since those will
 * not be read from connection and instead will be interpreted as another remote instruction, but
 * that instruction will essentially be garbage data and will cause the parsing loop to throw
 * exception and shutdown the connection which we do not want
```

summner commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ObjectTableAdvice.java:
```diff
@@ -0,0 +1,38 @@
+  public static void methodExit(
+      @Advice.Argument(0) final Object oe, @Advice.Return(readOnly = false) Target result) {
+
+    if (!DD_CONTEXT_CALL_ID.toString().equals(oe.toString())) {
```
While I'm not super happy about using toString in if in this case it avoids using reflection. Its also easily testable for all supported JDKs and is not supposed to change between patch releases.

Hmm actually I will add a comment explaining what object we actually are comparing here.

pawelchcki replied:
Whoops wrong account :)

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/RmiServerInstrumentation.java:
```diff
@@ -0,0 +1,105 @@
+
+      final AgentSpan span =
+          startSpan(callableContextStore)
+              .setTag(DDTags.RESOURCE_NAME, thiz.getClass().getSimpleName() + "#" + method)
```
Use decorator helper method here too.

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/RmiServerInstrumentation.java:
```diff
@@ -0,0 +1,105 @@
+      final AgentSpan.Context context = callableContextStore.get(Thread.currentThread());
+
+      if (context == null) {
+        return AgentTracer.startSpan("rmi.request");
```
AgentTracer methods were designed for static import.

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/ServerDecorator.java:
```diff
@@ -0,0 +1,23 @@
+import datadog.trace.agent.decorator.BaseDecorator;
+import datadog.trace.api.DDSpanTypes;
+
+public class ServerDecorator extends BaseDecorator {
```
Rename to `RmiServerDecorator` and extend `ServerDecorator`.

pawelchcki commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ObjectTableAdvice.java:
```diff
@@ -0,0 +1,38 @@
+  public static void methodExit(
+      @Advice.Argument(0) final Object oe, @Advice.Return(readOnly = false) Target result) {
+
+    if (!DD_CONTEXT_CALL_ID.toString().equals(oe.toString())) {
```
oops forgot to push a commit with the comment - added now

pawelchcki commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ContextDispatcher.java:
```diff
@@ -0,0 +1,45 @@
+      if (payload != null) {
+        final AgentSpan.Context context = propagate().extract(payload, GETTER);
+
+        callableContextStore.put(Thread.currentThread(), context);
```
If I use any of class injected by instrumentation there will be two - or more, instances of those classes.

When instrumenting both server instrumentation and context receiving call sites one will see instance 1, second will see instance 2. 

I think its because the RMI - soon before executing server implementation class sets 
Context class loader on current thread in (`Transport.java`):
```
    ClassLoader ccl = target.getContextClassLoader();

                ClassLoader savedCcl = Thread.currentThread().getContextClassLoader();

                try {
                    setContextClassLoader(ccl);
```

pawelchcki commented:
Thanks @tylerbenson for the review! Addressed all your comments. With this "thread context local" problem I'm open to alternative solutions since I too don't like the concept of injecting field into Thread object too much.

pawelchcki commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ObjectTableAdvice.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    final ContextStore<Thread, AgentSpan.Context> callableContextStore =
+        InstrumentationContext.get(Thread.class, AgentSpan.Context.class);
```
see: https://github.com/DataDog/dd-trace-java/pull/1128#discussion_r360355975

It could be possible but might require changes in the agent class injection

pawelchcki commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/StreamRemoteCallConstructorAdvice.java:
```diff
@@ -0,0 +1,143 @@
+  public static final ObjID ACTIVATOR_ID = new ObjID(ObjID.ACTIVATOR_ID);
+  public static final ObjID DGC_ID = new ObjID(ObjID.DGC_ID);
+  public static final ObjID REGISTRY_ID = new ObjID(ObjID.REGISTRY_ID);
+  public static final ObjID DD_CONTEXT_CALL_ID = new ObjID("Datadog.context_call.v2".hashCode());
```
there used to be v1 that i shipped to RMI. Just bumped it to 2. Now I reverted it to v1

pawelchcki commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/RmiContextInstrumentation.java:
```diff
@@ -0,0 +1,80 @@
+            safeHasSuperType(
+                named("sun.rmi.transport.StreamRemoteCall")
+                    .or(named("sun.rmi.transport.ObjectTable"))
+                    .or(named("sun.rmi.transport.ObjectEndpoint"))));
```
I think the current approach makes good logical distinctions between components (client, server and propagation). although splitting it into context.client and context.server does make a lot of sense. On it

pawelchcki commented:
Just noticed I missed few comments - on it.

pawelchcki commented:
Addressed last remaining comments

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/server/RmiServerContextInstrumentation.java:
```diff
@@ -0,0 +1,57 @@
+      "datadog.trace.instrumentation.rmi.context.ContextPayload",
+      "datadog.trace.instrumentation.rmi.context.ContextPropagator",
+      packageName + ".ContextDispatcher",
+      packageName + ".ObjectTableAdvice$NoopRemote"
```
Any time you have to inject an `Advice` class, that's a warning for me...  Can `NoopRemote` go somewhere else?

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ContextPropagator.java:
```diff
@@ -0,0 +1,103 @@
+  private static final ObjID REGISTRY_ID = new ObjID(ObjID.REGISTRY_ID);
+  public static final ObjID DD_CONTEXT_CALL_ID = new ObjID("Datadog.v1.context_call".hashCode());
+  private static final int CONTEXT_CHECK_CALL_OP_ID = -1;
+  public static final int CONTEXT_PASS_OPERATION_ID = -2;
```
This class, as well as these constants could use some docs.

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientAdvice.java:
```diff
@@ -0,0 +1,38 @@
+import java.lang.reflect.Method;
+import net.bytebuddy.asm.Advice;
+
+public class RmiClientAdvice {
```
Now that most of the logic is pulled out of the `Advice` classes, they're small enough they can be nested within the instrumentation class.  (This is completely optional and not very consistent as some instrumentation still keeps the advice class separate.)

tylerbenson replied:
You nested some other Advice classes, but not this one.  Was that intentional?

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/server/ContextDispatcher.java:
```diff
@@ -0,0 +1,47 @@
+import java.rmi.server.RemoteCall;
+import sun.rmi.server.Dispatcher;
+
+public class ContextDispatcher implements Dispatcher {
```
Could you add a class comment here?

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/context/ContextPayload.java:
```diff
@@ -0,0 +1,62 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ContextPayload implements Serializable {
```
Class comment?

pawelchcki commented:
Thanks @tylerbenson! Code is ready for next review

pawelchcki commented:
Thanks @tylerbenson - added last two fixes. Merging now! :)

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientDecorator.java:
```diff
@@ -0,0 +1,28 @@
+
+  @Override
+  protected String[] instrumentationNames() {
+    return new String[] {"rmi"};
```
This should include `rmi-client`.

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/server/RmiServerDecorator.java:
```diff
@@ -0,0 +1,23 @@
+
+  @Override
+  protected String[] instrumentationNames() {
+    return new String[] {"rmi"};
```
This should include `rmi-server`.

tylerbenson commented on dd-java-agent/instrumentation/rmi/src/main/java/datadog/trace/instrumentation/rmi/client/RmiClientDecorator.java:
```diff
@@ -0,0 +1,28 @@
+
+  @Override
+  protected String service() {
+    return "rmi";
```
Now that I think about it, we probably want this to be null since we expect the remote server to be traced and we don't want the extra hop in the service map.  @pawelchcki Can you submit a separate PR to fix this?

dougqh replied:
Yes, that probably sense in the usual case.  But isn't that a bit situational?  If they were calling a 3rd party service, then we'd probably prefer it as is.

I think is another example where our notion of peer service is slightly off.

pawelchcki replied:
If they were calling 3rd party service it would make sense to keep it. I think with RMI though chance of calling 3rd party service is slim :)
